### PR TITLE
Corrected type checking and propagation

### PR DESCRIPTION
**Pull request by kasper-piskorski**

### CODE REVIEW

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/Utility.java [hunk 1]
```diff
@@ -441,6 +441,15 @@ public static RoleType getNonMetaTopRole(RoleType role){
      * @return true if child is subtype of parent
      */
     public static boolean checkTypesCompatible(Type parent, Type child) {
-        return parent.equals(child) || parent.subTypes().contains(child);
+        return parent.subTypes().contains(child);
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/Utility.java]: Wouldn't it be more efficient to check if the super type's of `child` contains parent? That would involve serialising a lot less concepts.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/Utility.java [hunk 2]
```diff
@@ -441,6 +441,15 @@ public static RoleType getNonMetaTopRole(RoleType role){
...
+    }
+
+    /**
+     * @param parent type
+     * @param child type
+     * @return true if types do not belong to the same type hierarchy
+     */
+    public static boolean checkTypesDisjoint(Type parent, Type child) {
+        return !parent.subTypes().contains(child) && !child.subTypes().contains(parent);
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/Utility.java]: Ditto for this.

File: grakn-test/src/test/java/ai/grakn/test/graql/reasoner/QueryTest.java [hunk 3]
```diff
@@ -78,22 +80,23 @@ public void testCopyConstructor(){
 
     @Test
     public void testTwinPattern() {
```
Comment (fppt) [inline grakn-test/src/test/java/ai/grakn/test/graql/reasoner/QueryTest.java]: Could we get more details on this test? A comment or maybe change the name to be more descriptive.

File: grakn-test/src/test/java/ai/grakn/test/graql/reasoner/ReasoningTests.java [hunk 4]
```diff
@@ -312,13 +312,33 @@ public void reusingResources4() {
...
+                "$y isa entity1;" +
+                "(role1: $x, role2: $y) isa relation1;";
+        String queryString2 = queryString + "$y has name 'a';";
+        QueryAnswers answers = queryAnswers(qb.parse(queryString));
         assertEquals(answers.size(), 2);
+        QueryAnswers answers2 = queryAnswers(qb.parse(queryString2));
+        assertEquals(answers2.size(), 2);
+    }
+
+    @Test //Expected result: Single answer obtained only if the rule query containing super type is correctly executed.
```
Comment (fppt) [inline grakn-test/src/test/java/ai/grakn/test/graql/reasoner/ReasoningTests.java]: Putting these types of comments on the tests above would help a lot in debugging I think

**Comment by kasper-piskorski** [comment]
please retest this please - ai.grakn.test.engine.controller.TasksControllerTest.afterSendingTaskWithRunAt_ItIsDelayedInStorage

**Pull request merged**